inferelator_ng/bbsr_python.py

```diff
@@ -0,0 +1,316 @@
+            s.extend(ps)
+        print ('final s', len(s))
+        return s
+    else:
+        return None
+
+def BBSRforOneGene(ind, X, Y, pp, weights_mat, nS):
+    if ind % 100 == 0:
+        print('Progress: BBSR for gene', ind , '\n')
+
+    pp_i = pp.ix[ind,].values # converted to numpy array
+    pp_i_index = [l for l, j in enumerate(pp_i) if j]
+
+    if sum(pp_i) == 0:
+        return dict(ind=ind,pp=np.repeat(True, len(pp_i)).tolist(),betas=0, betas_resc=0)
+
+    # create BestSubsetRegression input
+    y = Y.ix[ind,:][:, np.newaxis]
+    x = X.ix[pp_i_index,:].transpose().values # converted to numpy array
+    g = np.matrix(weights_mat.ix[ind,pp_i_index],dtype=np.float)
+
+    # experimental stuff
+    spp = ReduceNumberOfPredictors(y, x, g, nS)
+
+    #check again
+    pp_i[pp_i==True] = spp # this could cause issues if they aren't the same length
+    pp_i_index = [l for l, j in enumerate(pp_i) if j]
+    x = X.ix[pp_i_index,:].transpose().values # converted to numpy array
+    g = np.matrix(weights_mat.ix[ind,pp_i_index],dtype=np.float)
+
+    betas = BestSubsetRegression(y, x, g)
+    betas_resc = PredErrRed(y, x, betas)
+
+    return (dict(ind=ind, pp=pp_i, betas=betas, betas_resc=betas_resc))
+
+
+
+def ReduceNumberOfPredictors(y, x, g, n):
+    K = x.shape[1] #what is the maximum size of K, print K
+    spp = None
+    if K <= n:
+        spp = np.repeat(True, K).tolist()
+        return spp
+
+    combos = np.hstack((np.diag(np.repeat(True,K)),CombCols(K)))
+    bics = ExpBICforAllCombos(y, x, g, combos)
+    bics_avg = np.sum(np.multiply(combos.transpose(),bics[:, np.newaxis]).transpose(),1)
+    bics_avg = list(bics_avg)
+    ret = np.repeat(False, K)
+    ret[np.argsort(bics_avg)[0:n]] = True
+    return ret
+
+
+def BestSubsetRegression(y, x, g):
+    # Do best subset regression by using all possible combinations of columns of
+    #x as predictors of y. Model selection criterion is BIC using results of
+    # Bayesian regression with Zellner's g-prior.
+    # Args:
+    #   y: dependent variable
+    #   x: independent variable
+    #   g: value for Zellner's g-prior; can be single value or vector
+    # Returns:
+    #   Beta vector of best mode
+    K = x.shape[1]
+    N = x.shape[0]
+    ret = []
+
+    combos = AllCombinations(K)
+    bics = ExpBICforAllCombos(y, x, g, combos)
+    not_done = True
+    while not_done:
+
+        best = np.argmin(bics)
+        betas = np.repeat(0.0,K)
+        if best > 0:
+
+            lst_combos_bool=combos[:, best]
+            lst_true_index = [i for i, j in enumerate(lst_combos_bool) if j]
+            x_tmp = x[:,lst_true_index]
+
+            try:
+                bhat = np.linalg.solve(np.dot(x_tmp.transpose(),x_tmp),np.dot(x_tmp.transpose(),y))
+                for m in xrange(len(lst_true_index)):
+                    ind_t=lst_true_index[m]
+                    betas[ind_t]=bhat[m]
+                not_done = False
+
+            except:
```

Comment:
What exception is expected here?  This unquailified except masked a NameError.  Please be more specific.